File: wultlibs/htmlreport/_Tab.py
# -*- coding: utf-8 -*-
# vim: ts=4 sw=4 tw=100 et ai si
#
# Copyright (C) 2019-2021 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
#
# Authors: Adam Hawley <adam.james.hawley@intel.com>

"""
This module defines what is expected by Jinja templates when adding a tab to HTML reports by
defining the 'Tab' dataclass.
"""

from dataclasses import dataclass
from typing import List, Dict

@dataclass
class Tab:
    """
    This class defines what is expected by the Jinja templates when adding a tab to the
    report.

    Jinja templates read from Tab objects to populate tabs in the report. Here is how it is done:
     1. The tab selector (the button you click to open a tab) is created with 'label' as the text in
        the button and 'id' as the HTML element ID.
     2. If 'Tab.tabs' is populated with child tabs, the template recursively adds these tabs.
     3. Depending on the 'category' of tab chosen, the template uses a different macro to populate
        the tab. The macro will be passed the dictionary 'mdata'.
    """

    # HTML tab element ID.
    id: str
    # Label for the tab selector.
    label: str
    # Child tabs (each child tab is of type 'Tab').
    tabs: List['Tab'] = None
    # If a 'category' is defined, it is used to populate the tab using the correct macro.
    # Possible values include 'metric', 'info' or None.
    category: str = None
    # Macros which populate the tab content will be provided the 'mdata' dictionary.
    mdata: Dict = None

File: js/src/tab.js
/*
 * -*- coding: utf-8 -*-
 * vim: ts=4 sw=4 tw=100 et ai si
 *
 * Copyright (C) 2021-2022 Intel, Inc.
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Author: Adam Hawley <adam.james.hawley@intel.com>
 */

import { LitElement } from 'lit'

/**
 * Generic class which handles visiblity of report tabs. When creating a new tab, it is recommended
 * to use a class which inherits from this class or create a new type. For example, if you want to
 * add a new 'data' tab, then use the '<sc-data-tab>' element over this one.
 *
 * @class ScTab
 * @extends {LitElement}
 */
export class ScTab extends LitElement {
    /**
     * Provides the template for when the tab is visible (active).
     */
    visibleTemplate () {
        throw new Error("Inherit from this class and implement 'visibleTemplate'.")
    }

    render () {
        return this.visibleTemplate()
    }
}

customElements.define('sc-tab', ScTab)
